eclipse-theia/theia
# Add support for compound launches

#### What it does
Fixes https://github.com/eclipse-theia/theia/issues/11302

- Extend debug model for compound launches
-- Support pre-launch task, stopAll, list of configurations
-- Properly parse compounds into configuration model
-- Support mapping to custom session option type
-- Support search of options by configuration, compound, and name

- Show compound launches in configuration dropdown
-- Query all available options
-- Use JSON (de-)serialization to allow multiple options with same name
--- Previously custom string was used for identification

- Extend session manager to start compound session options
- Ensure compound sessions can also be found from DebugMain

I tried to keep backwards compatibility with interfaces and parameters as good as possible.

#### How to test
- If you start Theia with Theia itself as a workspace, you should be able to see the already present compound launches `Launch Electron Backend & Frontend` and `Launch Browser Backend & Frontend`. But you can also add your own compound launches and play around with the options (stopAll, preLaunchTask).
- To test the call from within a plugin, I added a small plugin that has a few commands to trigger the existing launch configurations (with or without workspace) or provide a custom launch configuration: https://github.com/martin-fleck-at/theia/tree/issue-11302-test

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
